pkg/frontegg/scim_groups.go

```diff
@@ -0,0 +1,129 @@
+package frontegg
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/MaterializeInc/terraform-provider-materialize/pkg/clients"
+)
+
+const (
+	SCIMGroupsApiPathV1 = "/frontegg/identity/resources/groups/v1"
+)
+
+// ScimGroup represents the structure of a group in the response.
+type ScimGroup struct {
+	ID          string     `json:"id"`
+	Name        string     `json:"name"`
+	Description string     `json:"description"`
+	Metadata    string     `json:"metadata"`
+	Roles       []ScimRole `json:"roles"`
+	Users       []ScimUser `json:"users"`
+	ManagedBy   string     `json:"managedBy"`
+}
+
+// ScimRole represents the structure of a role within a group.
+type ScimRole struct {
+	ID          string `json:"id"`
+	Key         string `json:"key"`
+	Name        string `json:"name"`
+	Description string `json:"description"`
+	IsDefault   bool   `json:"is_default"`
+}
+
+// ScimUser represents the structure of a user within a group.
+type ScimUser struct {
+	ID    string `json:"id"`
+	Name  string `json:"name"`
+	Email string `json:"email"`
+}
+
+// SCIMGroupsResponse represents the overall structure of the response from the SCIM groups API.
+type SCIMGroupsResponse struct {
+	Groups []ScimGroup `json:"groups"`
+}
+
+// FetchSCIMGroups fetches all SCIM groups.
+func FetchSCIMGroups(ctx context.Context, client *clients.FronteggClient) ([]ScimGroup, error) {
+	endpoint := fmt.Sprintf("%s%s?_groupsRelations=rolesAndUsers", client.Endpoint, SCIMGroupsApiPathV1)
+	req, err := http.NewRequestWithContext(ctx, "GET", endpoint, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Add("Authorization", "Bearer "+client.Token)
+
+	resp, err := client.HTTPClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		var sb strings.Builder
+		_, err = io.Copy(&sb, resp.Body)
+		if err != nil {
+			return nil, err
+		}
+		return nil, fmt.Errorf("error reading SCIM groups: status %d, response: %s", resp.StatusCode, sb.String())
+	}
+
+	var groupsResponse SCIMGroupsResponse
+	if err := json.NewDecoder(resp.Body).Decode(&groupsResponse); err != nil {
+		return nil, err
+	}
+
+	return groupsResponse.Groups, nil
+}
+
+// Helper function to flatten the roles data
+func FlattenRoles(roles []ScimRole) []interface{} {
```

Comment:
Why are some functions capitalized and others not?

Reply:
Thanks Jun! 

If a function name starts with an uppercase letter, it will be exported to other packages. Eg. in this case the `FlattenRoles` func will be exposed to the `resources` and `datasources` packages if it is ever needed.